### ISSUE TITLE
[r/boards] initialize global AdminDAO

### ISSUE THREAD

**Issue by salmad3** [gnolang/gno#3345]
### Context:

The AdminDAO is a crucial component of the boards realm, responsible for managing high-level configurations and permissions. It serves as the primary administrative entity for the entire boards realm, separate from individual board management. The AdminDAO needs to be initialized with appropriate permissions and functionalities to oversee the boards realm effectively.

### Acceptance Criteria:

- [ ] Implements an AdminDAO struct with necessary fields:
    <details>
    <summary>Example</summary>

    ```go
    type AdminDAO struct {
        Members []Address
        Proposals []Proposal
        Config BoardsRealmConfig
    }

    type BoardsRealmConfig struct {
        MinBoardNameLength int
        DefaultBoardCreationCost uint64
        BoardNameValidationCallback func(string) (bool, uint64)
    }
    ```
    </details>

- [ ] An `InitializeAdminDAO` function sets up the initial state:
    <details>
    <summary>Example</summary>

    ```go
    func InitializeAdminDAO(initialMembers []Address) *AdminDAO {
        return &AdminDAO{
            Members: initialMembers,
            Proposals: []Proposal{},
            Config: BoardsRealmConfig{
                MinBoardNameLength: 3,
                DefaultBoardCreationCost: 1000000, // 1 GNOT
                BoardNameValidationCallback: defaultBoardNameValidation,
            },
        }
    }

    func defaultBoardNameValidation(name string) (bool, uint64) {
        if len(name) < 3 {
            return false, 0
        }
        if len(name) <= 6 {
            return true, 10000000 // 10 GNOT for short names
        }
        return true, 1000000 // 1 GNOT for regular names
    }
    ```
    </details>

- [ ] Implements a proposal system for AdminDAO decisions:
    - Create proposals
    - Vote on proposals
    - Execute approved proposals
- [ ] Includes a mechanism to add or remove AdminDAO members through proposals.
- [ ] Functions for the AdminDAO to manage global boards realm settings:
    - Set minimum board name length
    - Set default board creation cost
    - Set board name validation callback
- [ ] Ensures the AdminDAO can set and update the callback function for board name validation:
    <details>
    <summary>Example</summary>

    ```
    func (dao *AdminDAO) SetBoardNameValidationCallback(callback func(string) (bool, uint64)) {
        dao.Config.BoardNameValidationCallback = callback
    }
    ```
    </details>

- [ ] Includes a function for the AdminDAO to freeze a board in case of policy violations:
    <details>
    <summary>Example</summary>

    ```
    func (dao *AdminDAO) FreezeBoard(boardName string) error {
        // Implementation
    }
    ```
    </details>

- [ ]Includes a function for the AdminDAO to set realm-wide notifications:
    <details>
    <summary>Example</summary>

    ```
    func (dao *AdminDAO) SetRealmNotification(message string) {
        dao.Config.RealmNotification = message
    }
    ```
    </details>

- [ ] Ensures the AdminDAO can manage the upgrade process for the boards realm:
    - Set upgrade path
    - Approve upgrades
    - Manage transition to new versions

- [ ] Implements a mechanism to handle name change lookups:
    <details>
    <summary>Example</summary>

    ```
    type NameChangeRecord struct {
        OldName string
        NewName string
        Timestamp time.Time
    }

    func (dao *AdminDAO) RecordNameChange(oldName, newName string) {
        // Implementation
    }

    func (dao *AdminDAO) LookupCurrentName(name string) string {
        // Implementation
    }
    ```
    </details>

- [ ] Unit tests to verify the functionality of the AdminDAO
    - Member management
    - Proposal system
    - Configuration updates.

### Notes:

- The AdminDAO should reside within the boards realm, not as a separate realm.
- The AdminDAO should not have direct control over individual boards but should be able to intervene in extreme cases (e.g., freezing a board).
- The initial implementation should be simple, with the expectation that more complex governance mechanisms may be added in future versions.
- The AdminDAO should be designed with the possibility of future integration with a chain-level governance system (GovDAO) in mind.
- For the MVP, the AdminDAO will have no owner, only admins. The chain-level governance (GovDAO) will have the ability to replace admins in the future.
- The board name validation should initially enforce the same rules as the users realm (e.g., ending with three digits for free names).

**Comment by salmad3** [comment]
[Initial work](https://github.com/gnolang/gno/pull/3211) includes the following:

- An AdminDAO struct has been implemented, which includes members, proposals, and configuration fields.
- Functions for adding and removing members from the AdminDAO have been created.
- The AdminDAO has been integrated with the board creation process.
- Functions for inviting members and assigning roles have been implemented.
- A basic proposal system structure has been added to the AdminDAO (though it's noted that it needs further implementation).

**Comment by jeronimoalbi** [comment]
We are switching efforts to have an MVP by focusing in a default permissions implementation that doesn't use proposals for its functionality. Once the MVP is finished and once the voting mechanics are defined we will resume the work started in [feat/admindao-proposals](https://github.com/gnostudio/gno/tree/feat/admindao-proposals/examples/gno.land/p/demo/boards2/admindao) branch.

**Comment by salmad3** [comment]
**[08-01-2025]:**

- A top-level DAO governs the entire boards realm, holding ultimate authority (e.g., creating boards, managing upgrades).
- We'll launch with a simple owner/super role; additional roles and permissions can evolve in future versions.
- The "super" DAO logic will be separate from boards logic, based migration and maintainability benefits.
- The "super" DAO has discretion on memberships and flagging across all boards.
- The current approach lays groundwork for potential integration with other DAOs or new features (e.g., voting mechanisms) without major rewrites.

**Comment by moul** [comment]
It should not be prefixed as "boards." Commondao is not a subset of boards; rather, it is a dependency shared with other elements.

**Comment by salmad3** [comment]
@moul I suggest we keep this issue focused on the original AdminDAO criteria as-is, then extract it into its own realm under the CommonDAO model. We’ve already got some AdminDAO logic in the boards realm, so once we refactor, we can rename and move it cleanly.

**Comment by salmad3** [comment]
If this works, we'll open a separate issue for the CommonDAO initiative (not prefixed as "[r/boards]").

**Comment by jeronimoalbi** [comment]
PR #3558 contains an initial implementation of the Common DAO (ex Admin DAO). The implementation aims to be as simple as possible and to be somehow aligned with current DAO implementations.

After checking it ended up being quick to refactor `admindao` to be `commondao`, it will require some other changes in PR #3522 though.
